staging/src/k8s.io/kube-aggregator/pkg/util/dynamic_restconfig.go

```diff
@@ -0,0 +1,180 @@
+package util
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	"k8s.io/apiserver/pkg/server/egressselector"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const workItemKey = "key"
+
+type RestConfigProvider interface {
+	GetRestConfig() (*restclient.Config, error)
+	GetTransport() (http.RoundTripper, error)
+}
+
+// DynamicRestConfigProvider provides a restconfig and transport backed by dynamically loaded cert/key pair
+type DynamicRestConfigProvider struct {
```

Comment:
Thinking if we really need a controller, what would be an alternative design?

Reply:
yeah I am open to other ideas, I pretty much kept this in line with how are similar problems solved in apiserver (so the dynamiccertificates package)

Reply:
How about:

1. Adding a function to `APIAggregator` that would return a `cert` and a `key`
2. Adding a new method to `APIAggregator` that would call `updateAPIService` for all `proxyHandlers` and would use a func from `1`.
3. Wire  `Notify` method to call the new function from `2`

Alternatively:
1. Adding a new method to `APIAggregator` that would call `updateAPIService` for all `proxyHandlers` and would accept a cert and a key
2. Wire  `Notify` method to call the new function from `1` and providing a cert and a key

Doing this that way we don't modify the hot path and we take potential changes in services into account. It should work if the old cert is still valid. What do you think?

@sttts FYI

Reply:
@p0lyn0mial that solves it for proxy_handlers, and then available_controller will have to implement the exact same thing (that's why I went the way of extracting that logic outside, so both places can reuse it). 

If we go and extract it then the only difference between your proposal and current implementation is whether it uses the workqueue or not. I like reusing workqueue since it's already proven piece of code that solves the async update for me.

No hard opinion here though, just putting out here why I went the way I went. I am totally fine shifting the direction :-)

Reply:
It seems that `available_controller` only needs a `restConfig` - slightly different one than `proxy_handlers`. I think that before it didn't use `MTLS`, now it does. I don't know why it was like that before but it should work with `MTLS` as well, right?

Reply:
alright, let's try with the controller, left some comments.

Reply:
it does not use mtls as I pass insecure=true here https://github.com/kubernetes/kubernetes/pull/88120/files#diff-d92af716cdd9489e3f57d61b53adad72R216 ... should I make that more obvious somehow?

Reply:
I think I got the confusion around "controller" now, the Run method I think is definitely something we don't need, it was even unused. I removed it, so now it really just asynchronously processes notifications.